File: src/recursion_and_dynamic_programming/OneAwayLCCI.java
package recursion_and_dynamic_programming;

/**
 * @Author: Wenhang Chen
 * @Description:字符串有三种编辑操作:插入一个字符、删除一个字符或者替换一个字符。 给定两个字符串，编写一个函数判定它们是否只需要一次(或者零次)编辑。
 * <p>
 *  
 * <p>
 * 示例 1:
 * <p>
 * 输入:
 * first = "pale"
 * second = "ple"
 * 输出: True
 *  
 * <p>
 * 示例 2:
 * <p>
 * 输入:
 * first = "pales"
 * second = "pal"
 * 输出: False
 * @Date: Created in 9:06 3/18/2020
 * @Modified by:
 */
public class OneAwayLCCI {
    public boolean oneEditAway(String first, String second) {
        if (first == null && second == null) return true;
        if (first == null || second == null) return false;
        if (first.equals(second)) return true;
        int cnt = 0;
        int m = first.length();
        int n = second.length();
        if (m == n) {
            for (int i = 0; i < m; i++) {
                if (first.charAt(i) != second.charAt(i)) cnt++;
            }
        } else if (m - n == 1) {
            int i = 0;
            int j = 0;
            while (i < m && j < n) {
                if (first.charAt(i) == second.charAt(j)) {
                    i++;
                    j++;
                } else {
                    i++;
                    cnt++;
                }
            }
        } else if (n - m == 1) {
            int i = 0;
            int j = 0;
            while (i < m && j < n) {
                if (first.charAt(i) == second.charAt(j)) {
                    i++;
                    j++;
                } else {
                    j++;
                    cnt++;
                }
            }
        } else {
            return false;
        }
        return cnt <= 1;
    }
}

File: src/leetcode/one_away_lCCI.py
class Solution:
    def oneEditAway(self, first: str, second: str) -> bool:
        dp = [[0] * (len(first) + 1) for _ in range(len(second) + 1)]
        n1 = len(first) + 1
        n2 = len(second) + 1
        for i in range(1, n1):
            dp[0][i] = dp[0][i - 1] + 1
        for i in range(1, n2):
            dp[i][0] = dp[i - 1][0] + 1
        for i in range(1, n2):
            for j in range(1, n1):
                if first[j - 1] == second[i - 1]:
                    dp[i][j] = dp[i - 1][j - 1]
                else:
                    dp[i][j] = min(dp[i - 1][j - 1], dp[i][j - 1], dp[i - 1][j]) + 1
        # print(dp)
        return dp[n2 - 1][n1 - 1] <= 1


sol = Solution()
sol.oneEditAway("pale", "ple")
